// foundry-local-sdk/dist/foundryLocalManager.js
import { Configuration } from './configuration.js';
import { CoreInterop } from './detail/coreInterop.js';
import { ModelLoadManager } from './detail/modelLoadManager.js';
import { Catalog } from './catalog.js';
/**
 * The main entry point for the Foundry Local SDK.
 * Manages the initialization of the core system and provides access to the Catalog and ModelLoadManager.
 */
export class FoundryLocalManager {
    static instance;
    config;
    coreInterop;
    _modelLoadManager;
    _catalog;
    _urls = [];
    constructor(config) {
        this.config = config;
        this.coreInterop = new CoreInterop(this.config);
        this.coreInterop.executeCommand("initialize", { Params: this.config.params });
        this._modelLoadManager = new ModelLoadManager(this.coreInterop);
        this._catalog = new Catalog(this.coreInterop, this._modelLoadManager);
    }
    /**
     * Creates the FoundryLocalManager singleton with the provided configuration.
     * @param config - The configuration settings for the SDK (plain object).
     * @returns The initialized FoundryLocalManager instance.
     * @example
     * ```typescript
     * const manager = FoundryLocalManager.create({
     *   appName: 'MyApp',
     *   logLevel: 'info'
     * });
     * ```
     */
    static create(config) {
        if (!FoundryLocalManager.instance) {
            const internalConfig = new Configuration(config);
            FoundryLocalManager.instance = new FoundryLocalManager(internalConfig);
        }
        return FoundryLocalManager.instance;
    }
    /**
     * Gets the Catalog instance for discovering and managing models.
     * @returns The Catalog instance.
     */
    get catalog() {
        return this._catalog;
    }
    /**
     * Gets the URLs where the web service is listening.
     * Returns an empty array if the web service is not running.
     * @returns An array of URLs.
     */
    get urls() {
        return this._urls;
    }
    /**
     * Starts the local web service.
     * Use the `urls` property to retrieve the bound addresses after the service has started.
     * If no listener address is configured, the service defaults to `127.0.0.1:0` (binding to a random ephemeral port).
     * @throws Error - If starting the service fails.
     */
    startWebService() {
        const response = this.coreInterop.executeCommand("start_service");
        try {
            this._urls = JSON.parse(response);
        }
        catch (error) {
            throw new Error(`Failed to decode JSON response from start_service: ${error}. Response was: ${response}`);
        }
    }
    /**
     * Stops the local web service.
     * @throws Error - If stopping the service fails.
     */
    stopWebService() {
        if (this._urls.length > 0) {
            this.coreInterop.executeCommand("stop_service");
            this._urls = [];
        }
    }
}
