// foundry-local-sdk/dist/modelVariant.js
import { ChatClient } from './openai/chatClient.js';
import { AudioClient } from './openai/audioClient.js';
/**
 * Represents a specific variant of a model (e.g., a specific quantization or format).
 * Contains the low-level implementation for interacting with the model.
 */
export class ModelVariant {
    _modelInfo;
    coreInterop;
    modelLoadManager;
    constructor(modelInfo, coreInterop, modelLoadManager) {
        this._modelInfo = modelInfo;
        this.coreInterop = coreInterop;
        this.modelLoadManager = modelLoadManager;
    }
    /**
     * Gets the unique identifier of the model variant.
     * @returns The model ID.
     */
    get id() {
        return this._modelInfo.id;
    }
    /**
     * Gets the alias of the model.
     * @returns The model alias.
     */
    get alias() {
        return this._modelInfo.alias;
    }
    /**
     * Gets the detailed information about the model variant.
     * @returns The ModelInfo object.
     */
    get modelInfo() {
        return this._modelInfo;
    }
    /**
     * Checks if the model variant is cached locally.
     * @returns True if cached, false otherwise.
     */
    get isCached() {
        const cachedModels = JSON.parse(this.coreInterop.executeCommand("get_cached_models"));
        return cachedModels.includes(this._modelInfo.id);
    }
    /**
     * Checks if the model variant is loaded in memory.
     * @returns True if loaded, false otherwise.
     */
    async isLoaded() {
        const loadedModels = await this.modelLoadManager.listLoaded();
        return loadedModels.includes(this._modelInfo.id);
    }
    /**
     * Downloads the model variant.
     * @param progressCallback - Optional callback to report download progress.
     * @throws Error - If progress callback is provided (not implemented).
     */
    async download(progressCallback) {
        const request = { Params: { Model: this._modelInfo.id } };
        if (!progressCallback) {
            this.coreInterop.executeCommand("download_model", request);
        }
        else {
            await this.coreInterop.executeCommandStreaming("download_model", request, (chunk) => {
                const progress = parseFloat(chunk);
                if (!isNaN(progress)) {
                    progressCallback(progress);
                }
            });
        }
    }
    /**
     * Gets the local file path of the model variant.
     * @returns The local file path.
     */
    get path() {
        const request = { Params: { Model: this._modelInfo.id } };
        return this.coreInterop.executeCommand("get_model_path", request);
    }
    /**
     * Loads the model variant into memory.
     * @returns A promise that resolves when the model is loaded.
     */
    async load() {
        await this.modelLoadManager.load(this._modelInfo.id);
    }
    /**
     * Removes the model variant from the local cache.
     */
    removeFromCache() {
        this.coreInterop.executeCommand("remove_cached_model", { Params: { Model: this._modelInfo.id } });
    }
    /**
     * Unloads the model variant from memory.
     * @returns A promise that resolves when the model is unloaded.
     */
    async unload() {
        await this.modelLoadManager.unload(this._modelInfo.id);
    }
    /**
     * Creates a ChatClient for interacting with the model via chat completions.
     * @returns A ChatClient instance.
     */
    createChatClient() {
        return new ChatClient(this._modelInfo.id, this.coreInterop);
    }
    /**
     * Creates an AudioClient for interacting with the model via audio operations.
     * @returns An AudioClient instance.
     */
    createAudioClient() {
        return new AudioClient(this._modelInfo.id, this.coreInterop);
    }
}
