// foundry-local-sdk/dist/model.js
/**
 * Represents a high-level AI model that may have multiple variants (e.g., quantized versions, different formats).
 * Manages the selection and interaction with a specific model variant.
 */
export class Model {
    _alias;
    _variants;
    selectedVariant;
    constructor(variant) {
        this._alias = variant.alias;
        this._variants = [variant];
        this.selectedVariant = variant;
    }
    /**
     * Adds a new variant to this model.
     * Automatically selects the new variant if it is cached and the current one is not.
     * @param variant - The model variant to add.
     * @throws Error - If the variant's alias does not match the model's alias.
     */
    addVariant(variant) {
        if (variant.alias !== this._alias) {
            throw new Error("Variant alias does not match model alias.");
        }
        this._variants.push(variant);
        // prefer the highest priority locally cached variant
        if (variant.isCached && !this.selectedVariant.isCached) {
            this.selectedVariant = variant;
        }
    }
    /**
     * Selects a specific variant by its ID.
     * @param modelId - The ID of the variant to select.
     * @throws Error - If the variant with the specified ID is not found.
     */
    selectVariant(modelId) {
        for (const variant of this._variants) {
            if (variant.id === modelId) {
                this.selectedVariant = variant;
                return;
            }
        }
        throw new Error(`Model variant with id ${modelId} not found.`);
    }
    /**
     * Gets the ID of the currently selected variant.
     * @returns The ID of the selected variant.
     */
    get id() {
        return this.selectedVariant.id;
    }
    /**
     * Gets the alias of the model.
     * @returns The model alias.
     */
    get alias() {
        return this._alias;
    }
    /**
     * Checks if the currently selected variant is cached locally.
     * @returns True if cached, false otherwise.
     */
    get isCached() {
        return this.selectedVariant.isCached;
    }
    /**
     * Checks if the currently selected variant is loaded in memory.
     * @returns True if loaded, false otherwise.
     */
    async isLoaded() {
        return await this.selectedVariant.isLoaded();
    }
    /**
     * Gets all available variants for this model.
     * @returns An array of ModelVariant objects.
     */
    get variants() {
        return this._variants;
    }
    /**
     * Downloads the currently selected variant.
     * @param progressCallback - Optional callback to report download progress.
     */
    download(progressCallback) {
        return this.selectedVariant.download(progressCallback);
    }
    /**
     * Gets the local file path of the currently selected variant.
     * @returns The local file path.
     */
    get path() {
        return this.selectedVariant.path;
    }
    /**
     * Loads the currently selected variant into memory.
     * @returns A promise that resolves when the model is loaded.
     */
    async load() {
        await this.selectedVariant.load();
    }
    /**
     * Removes the currently selected variant from the local cache.
     */
    removeFromCache() {
        this.selectedVariant.removeFromCache();
    }
    /**
     * Unloads the currently selected variant from memory.
     * @returns A promise that resolves when the model is unloaded.
     */
    async unload() {
        await this.selectedVariant.unload();
    }
    /**
     * Creates a ChatClient for interacting with the model via chat completions.
     * @returns A ChatClient instance.
     */
    createChatClient() {
        return this.selectedVariant.createChatClient();
    }
    /**
     * Creates an AudioClient for interacting with the model via audio operations.
     * @returns An AudioClient instance.
     */
    createAudioClient() {
        return this.selectedVariant.createAudioClient();
    }
}
